### PR TITLE
1D inverse problems showcase from hackathon

### PR DESCRIPTION
**Pull request by billlion**
## Checklist when you are ready to request a review

- [x] I have performed a self-review of my code
- [x] I have created a new folder, containing my contributions which are in the form of jupyter notebooks, with any necessary supporting python files, and a LICENSE file.
- [x] I have added a description of my contribution(s) to the top of my file(s)
- [x] If publicly available, I have added a link to the dataset used near the top of my file(s)
- [x] I have added the CIL version I ran with near the top of my file(s)
- [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL-User-Showcase.
- [x] I confirm that the contribution does not violate any intellectual property rights of third parties.
- [x] I confirm that I have added license headers to all of the files I am contributing (with a license of my choice)
- [x] Change pull request label to 'Waiting for review' 

Note: for an example of a contribution, where a license header, description, data link and CIL version has been added, please
see: [example_contribution](https://github.com/TomographicImaging/CIL-User-Showcase/blob/main/example_contribution)

### CODE REVIEW

**Review by lauramurgatroyd — CHANGES_REQUESTED** [review]
Hi Bill,

Thanks for the contribution!

Just some initial comments on the format of the notebook.

**Review by jakobsj — CHANGES_REQUESTED** [review]
Hi @billlion many thanks for adding this nice notebook of a 1D case.

In addition to the comments by @lauramurgatroyd relating to general tidying up, removing empty cells, old comments and unused code etc., I have the following suggestions/questions on the content/flow of the notebook:

- Just below the notebook title, please add your name and institution for author.
- The loglog plot of the singular value has a steep jump at the end. Is this to be expected? Is the problem numerically rank-deficient?
- Just before that, the comment "expected for an operator that approximates the inverse of two derivatives" is not quite clear to me.  Could you perhaps explain the deriv2 problem a bit more, that would add helpful context for those who may not know Regularization Tools and/or the deriv2 problem, and then elaborate on what is meant by "two derivatives".
- Please provide the notebook with outputs run once in order from a fresh kernel.
- The result of the first tikhonov (using lstsq and numpy) looks quite different from the true x. Is this expected? Is there a bug? Please add some commentary to the notebook to explain what we observe and how that fits with the theory.
- When the notebook turns to CIL types, the notebook has limited commentary. Could you add some more to guide the reader? For example, explain why we run CGLS, what we see from the CGLS solution (which is very wavy, is that expected?)
- After the first part on CIL, the notebook returns to look at numpy A and bn? What is the point there? Could this be moved up to before the CIL stuff starts to have all the numpy stuff together? Also the xlq appears to plotted twice, did you mean to make a comparison with something else?
- The last part with Tikhonov in CIL is nice but lacks explanation. Would it make sense to compare the CIL Tikhonov with the numpy tikhonov and not just the true x?

Once again, thank you for adding this example. With a bit of tidying up, restructuring, checking the deriv2 implementation and additional commentary, it will make a very nice notebook showcasing how easy it is to set up a 1D problem in CIL, so that users can explore classic 1D inverse problems and at the same time be set up for larger and higher-dimensional inverse problems.

**Comment by MargaretDuff** [comment]
I created a new pull request to https://github.com/billlion/CIL-User-Showcase that dealt with some of the issues

File: 1DIP/deriv2_cgls.ipynb [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7f0222cb",
+   "metadata": {},
+   "source": [
+    "This notebook was developed as part of the CCPi CIL Hackathon https://ccpi.ac.uk/events/byod-cil-hackathon/ in March 2023 in Cambridge as part of the Rich Nonlinear Tomography programme at the Isaac Newton Institute for Mathematical Sciences https://www.newton.ac.uk/event/rnt/.  The CIL is suppored by the CCPi EPSRC grant EP/T026677/1 and the Isaac NewtonInstitute by EP/R014604/1 The author would like to thank the Isaac Newton Institute for support  and hospitality.(c) W.R.B. Lionheart 2023. Apache License"
```
[the remark this answers is not in the document]
Comment (MargaretDuff) [reply 1DIP/deriv2_cgls.ipynb]: Fixed

File: 1DIP/deriv2_cgls.ipynb [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+      "text/plain": [
+       "<Figure size 640x480 with 1 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "import numpy as np\n",
```
[the remark this answers is not in the document]
Comment (MargaretDuff) [reply 1DIP/deriv2_cgls.ipynb:55]: Fixed

File: 1DIP/deriv2_cgls.ipynb [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
...gUVEREScngKLiIiIOD0FFhEREXF6CiwiIiLi9DzNLqAs2O12jh8/jr+/PxaLxexyRERE5AIYhkFGRgahoaFYrecfQ3GLwHL8+HHCwsLMLkNEREQuwpEjR6hfv/55z3GLwOLv7w84PnBAQIDJ1YiIiMiFSE9PJywsrPj3+Pm4RWA50wYKCAhQYBEREXExFzKdQ5NuRURExOkpsIiIiIjTU2ARERERp6fAIiIiIk7vogLL1KlTCQ8Px9fXl8jISNauXXvOc7dv307//v0JDw/HYrEwefLkS35NERERqVxKHVjmzZvHmDFjmDhxIhs2bKBDhw706dOHpKSks56fnZ1N48aNeeWVVwgJCSmT1xQREZHKxWIYhlGaJ0RGRtK1a1emTJkCOHaZDQsL45FHHmHs2LHnfW54eDijR49m9OjRZfaa4FjHHRgYSFpampY1i4iIuIjS/P4u1QhLfn4+cXFxREdH//ECVivR0dHExsZeVLEX85p5eXmkp6eX+BIRERH3VarAkpycjM1mIzg4uMTjwcHBJCQkXFQBF/OakyZNIjAwsPhL2/KLiIi4N5dcJTRu3DjS0tKKv44cOWJ2SSIiIlKOSrU1f61atfDw8CAxMbHE44mJieecUFser+nj44OPj89FvZ+IiIi4nlKNsHh7e9OlSxeWLVtW/JjdbmfZsmVERUVdVAHl8ZoiIiLiXkp988MxY8YQExNDREQE3bp1Y/LkyWRlZTFkyBAABg8eTL169Zg0aRLgmFS7Y8eO4v9/7NgxNm3aRLVq1WjatOkFvaaIiIiYxG6D5a+Ahxf0ftq0MkodWAYMGMDJkyeZMGECCQkJdOzYkSVLlhRPmo2Pj8dq/WPg5vjx43Tq1Kn4+LXXXuO1116jd+/eLF++/IJeU0REREyQkQBf3g+HfgMs0Lof1G5uSiml3ofFGWkfFhERkTK2/2dY8ABknQTvanDjZGh/R5m+RWl+f5d6hEVERETcmK0QVrwCv74GGBDcFu74EGo1M7UsBRYRERFxSD/haAEdXuk47jIErpsEXn7m1oUCi4iIiADsW+ZoAWUnO1pAN70J7W43u6piCiwiIiKVma0Qlk+C317H0QJqV9QCamp2ZSUosIiIiFRW6cfhi2EQv8pxHDEU+kwCL19z6zoLBRYREZHKaN9PRS2gU07ZAvorBRYREZHKxFYIv7wMK99wHDtp
```
[the remark this answers is not in the document]
Comment (MargaretDuff) [reply 1DIP/deriv2_cgls.ipynb]: Fixed

**Comment by MargaretDuff** [comment]
Reviewed by Jakob and the pull request on https://github.com/billlion/CIL-User-Showcase  should be ready to merge

**Review by jakobsj — APPROVED** [review]
I am happy to see this updated version merged, thanks all!

**Comment by jakobsj** [comment]
@lauramurgatroyd Merging awaits resolving of the changes you requested - could you take a look and see if you are happy with merging this version please?

File: 003_1D_integral_inverse_problem/__pycache__/deriv2.cpython-310.pyc
Comment (lauramurgatroyd) [inline 003_1D_integral_inverse_problem/__pycache__/deriv2.cpython-310.pyc:1]: Is this file needed?

File: 003_1D_integral_inverse_problem/deriv2.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import numpy as np
```
Comment (lauramurgatroyd) [inline 003_1D_integral_inverse_problem/deriv2.py:1]: I think we should have a license header on this file too

File: 003_1D_integral_inverse_problem/deriv2_cgls.ipynb [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+    "#      http://www.apache.org/licenses/LICENSE-2.0\n",
+    "#\n",
+    "#  Unless required by applicable law or agreed to in writing, software\n",
+    "#  distributed under the License is distributed on an \"AS IS\" BASIS,\n",
+    "#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
+    "#  See the License for the specific language governing permissions and\n",
+    "#  limitations under the License.\n",
+    "#\n",
+    "#   Authored by:  Bill Lionheart (University of Manchester),\n",
+    "#   Edited by:  Margaret DUff (STFC - UKRI)"
```
Comment (lauramurgatroyd) [inline 003_1D_integral_inverse_problem/deriv2_cgls.ipynb:26]:
```suggestion
    "#   Edited by:  Margaret Duff (STFC - UKRI)"
```

**Review by lauramurgatroyd — CHANGES_REQUESTED** [review]
Sorry I just noticed a couple of extra things when I came back to resolve comment threads

**Comment by MargaretDuff** [comment]
Thanks @lauramurgatroyd! @billlion I created a new pull request to deal with Laura's comments - if you could approve that would be great!

**Comment by MargaretDuff** [comment]
Thanks @billlion! @lauramurgatroyd - could you check you are happy and merge if so?

**Review by lauramurgatroyd — APPROVED** [review]

**Pull request merged**